File: src/classes/LspbTrajPlanner.m
classdef LspbTrajPlanner < handle
    % lspb trajectory(trapezoidal velocity profile)
    % TO DO: trajectory through a sequence of points
    
    properties
        vmax
        amax
        
        np
        t0
        tf
        v0
        vf
        q0
        qf
        ta
        td
        dir
        maxvel_reached
    end
    
    methods
        function obj = LspbTrajPlanner(pos, max_vel, max_acc, duration, vel_cons)
            % assume that: jmin = -jmax, amin = -amax, vmin = -vmax, t0 = 0
            % generic initial and final values of velocity
            % initial and final accelerations set to zeros
            obj.np = length(pos);
            obj.dir = sign(pos(2)-pos(1));
            h = abs(pos(2)-pos(1));
            if h>max_vel^2/max_acc
                obj.maxvel_reached = 1;
            else
                obj.maxvel_reached = 0;
            end
            if nargin==3
                obj.TransformPVA(pos, [0,0], max_vel, max_acc);
                obj.SetNoTimeLimit(h);
            elseif nargin==4
                obj.TransformPVA(pos, [0,0], max_vel, max_acc);
                obj.SetTimeLimit(h, duration);
            elseif nargin==5
                obj.TransformPVA(pos, vel_cons, max_vel, max_acc);
                obj.SetVelConstraint(h);
            end
        end

        function TransformPVA(obj, pos, vel, max_vel, max_acc)
            obj.q0 = pos(1)*obj.dir;
            obj.qf = pos(2)*obj.dir;
            obj.v0 = vel(1)*obj.dir;
            obj.vf = vel(2)*obj.dir;
            obj.vmax = 0.5*(obj.dir+1)*max_vel-0.5*(obj.dir-1)*max_vel;
            obj.amax = 0.5*(obj.dir+1)*max_acc-0.5*(obj.dir-1)*max_acc;
        end

        function SetNoTimeLimit(obj, h)
            obj.t0 = 0;
            if obj.maxvel_reached
                obj.ta = obj.vmax/obj.amax;
                obj.tf = h/obj.vmax+obj.vmax/obj.amax;
            else
                obj.ta = sqrt(h/obj.amax);
                obj.tf = 2*obj.ta;
                obj.vmax = h/obj.ta;
            end
            obj.td=obj.ta;
        end

        function SetTimeLimit(obj, h, duration)
            if isscalar(duration)
                obj.t0 = 0; obj.tf = duration;
            else
                obj.t0 = duration(1); obj.tf = duration(end);
            end
            time_length = obj.tf-obj.t0;
            if time_length>=h/obj.vmax+obj.vmax/obj.amax
                if obj.maxvel_reached
                    a = 1; b = -time_length*obj.amax; c = h*obj.amax;
                    obj.vmax = (-b-sqrt(b^2-4*a*c))/2/a;
                    obj.ta = obj.vmax/obj.amax;
                else
                    obj.vmax = 2*h/time_length;
                    obj.ta = 0.5*time_length;
                    obj.amax = obj.vmax/obj.ta;
                end
                obj.td=obj.ta;
            else
                error('input a larger time')
            end
        end

        function SetVelConstraint(obj, h)
            obj.t0 = 0;
            if obj.amax*h<0.5*abs(obj.v0^2-obj.vf^2)
                error('the trajectory does not exit');
            else
                if h*obj.amax>(obj.vmax^2-0.5*(obj.v0^2+obj.vf^2))
                    obj.maxvel_reached = 1;
                    obj.ta = (obj.vmax-obj.v0)/obj.amax;
                    obj.td = (obj.vmax-obj.vf)/obj.amax;
                    obj.tf = h/obj.vmax+0.5*obj.vmax/obj.amax*(1-obj.v0/obj.vmax)^2 ...
                            +0.5*obj.vmax/obj.amax*(1-obj.vf/obj.vmax)^2;
                else
                    obj.maxvel_reached = 0;
                    obj.vmax = sqrt(h*obj.amax+0.5*(obj.v0^2+obj.vf^2));
                    obj.ta = (obj.vmax-obj.v0)/obj.amax;
                    obj.td = (obj.vmax-obj.vf)/obj.amax;
                    obj.tf = obj.ta+obj.td;
                end
            end
        end
        
        function [pos, vel, acc] = GenerateTraj(obj, dt)
            pos = []; vel = []; acc = [];
            time_pnts = (obj.tf-obj.t0)/dt;
            for idx = 0:time_pnts
                t = obj.t0+dt*idx;
                [p, v, a] = obj.GenerateMotion(t);
                pos = [pos, p];
                vel = [vel, v];
                acc = [acc, a];
            end
        end
        
        function [p, v, a] = GenerateMotion(obj, t)
%             vm = obj.vmax*obj.dir;
%             am = obj.amax*obj.dir;
            if obj.maxvel_reached
                if t>=obj.t0 && t<obj.ta+obj.t0
                    p = obj.q0+obj.v0*(t-obj.t0)+0.5*(obj.vmax-obj.v0)/obj.ta*(t-obj.t0)^2;
                    v = obj.v0+(obj.vmax-obj.v0)/obj.ta*(t-obj.t0);
                    a = obj.amax;
                elseif t>=obj.ta+obj.t0 && t<obj.tf-obj.td
                    p = obj.q0+0.5*obj.v0*obj.ta+obj.vmax*(t-obj.t0-0.5*obj.ta);
                    v = obj.vmax;
                    a = 0;
                elseif t>=obj.tf-obj.td && t<=obj.tf
                    p = obj.qf-obj.vf*(obj.tf-t)-0.5*(obj.vmax-obj.vf)/obj.td*(obj.tf-t)^2;
                    v = obj.vf+(obj.vmax-obj.vf)/obj.td*(obj.tf-t);
                    a = -obj.amax;
                end
            else
                if t>=obj.t0 && t<=obj.t0+obj.ta
                    p = obj.q0+obj.v0*(t-obj.t0)+0.5*(obj.vmax-obj.v0)/obj.ta*(t-obj.t0)^2;
                    v = obj.v0+(obj.vmax-obj.v0)/obj.ta*(t-obj.t0);
                    a = obj.amax;
                elseif t> obj.tf-obj.td && t<=obj.tf
                    p = obj.qf-obj.vf*(obj.tf-t)-0.5*(obj.vmax-obj.vf)/obj.td*(obj.tf-t)^2;
                    v = obj.vf+(obj.vmax-obj.vf)/obj.td*(obj.tf-t);
                    a = -obj.amax;
                end
            end
            p = p*obj.dir; v = v*obj.dir; a = a*obj.dir;
        end
        
        function PlotAVP(obj, dt)
            [q, dq, ddq] = obj.GenerateTraj(dt);
            t = obj.t0:dt:obj.tf;
            subplot(3,1,1); plot(t, q); grid on; ylabel('position');     
            subplot(3,1,2); plot(t, dq); grid on; ylabel('velocity');
            subplot(3,1,3); plot(t, ddq); grid on; ylabel('acceleration');
        end
        
        
    end
end

File: src/cfiles/LspbTrajPlanner.cpp
#include "LspbTrajPlanner.h"


LspbTrajPlanner::LspbTrajPlanner(Vector2d pos, double max_vel, double max_acc)
{
	InitPlanner(pos, max_vel, max_acc);
}

LspbTrajPlanner::LspbTrajPlanner(Vector2d pos, double max_vel, double max_acc, double tf)
{
	InitPlanner(pos, max_vel, max_acc, tf);
}

LspbTrajPlanner::LspbTrajPlanner(Vector2d pos, double max_vel, double max_acc, Vector2d duration)
{
	InitPlanner(pos, max_vel, max_acc, duration);
}

LspbTrajPlanner::LspbTrajPlanner(Vector2d pos, double max_vel, double max_acc, Vector2d duration, Vector2d vel_con)
{
	InitPlanner(pos, max_vel, max_acc, duration, vel_con);
}

void LspbTrajPlanner::InitPlanner(Vector2d pos, double max_vel, double max_acc, Vector2d duration, Vector2d vel_con)
{
	_np = (int)pos.size();
	_dir = MathTools::Sign(pos(1)-pos(0));
	double h = fabs(pos(1)-pos(0));
	if (h>pow(max_vel, 2)/max_acc)
		_maxvel_reached = true;
	else
		_maxvel_reached = false;
	TransformPVA(pos, vel_con, max_vel, max_acc);
	SetVelConstraint(h);
}

void LspbTrajPlanner::InitPlanner(Vector2d pos, double max_vel, double max_acc, double tf)
{
	_np = (int)pos.size();
	_dir = MathTools::Sign(pos(1)-pos(0));
	double h = fabs(pos(1)-pos(0));
	if (h>pow(max_vel, 2)/max_acc)
		_maxvel_reached = true;
	else
		_maxvel_reached = false;
	Vector2d vel_con(0, 0);
	TransformPVA(pos, vel_con, max_vel, max_acc);
	Vector2d duration(0, tf);
	SetTimeLimit(h, duration);
}

void LspbTrajPlanner::InitPlanner(Vector2d pos, double max_vel, double max_acc, Vector2d duration)
{
	_np = (int)pos.size();
	_dir = MathTools::Sign(pos(1)-pos(0));
	double h = fabs(pos(1)-pos(0));
	if (h>pow(max_vel, 2)/max_acc)
		_maxvel_reached = true;
	else
		_maxvel_reached = false;
	Vector2d vel_con(0, 0);
	TransformPVA(pos, vel_con, max_vel, max_acc);
	SetTimeLimit(h, duration);
}

void LspbTrajPlanner::InitPlanner(Vector2d pos, double max_vel, double max_acc)
{
	_np = (int)pos.size();
	_dir = MathTools::Sign(pos(1)-pos(0));
	double h = fabs(pos(1)-pos(0));
	if (h>pow(max_vel, 2)/max_acc)
		_maxvel_reached = true;
	else
		_maxvel_reached = false;
	Vector2d vel_con(0, 0);
	TransformPVA(pos, vel_con, max_vel, max_acc);
	SetNoTimeLimit(h);
}

void LspbTrajPlanner::TransformPVA(Vector2d pos, Vector2d vel, double max_vel, double max_acc)
{
	_q0 = pos(0)*_dir;
	_qf = pos(1)*_dir;
	_v0 = vel(0)*_dir;
	_vf = vel(1)*_dir;
	_vmax = 0.5*(_dir+1)*max_vel-0.5*(_dir-1)*max_vel;
	_amax = 0.5*(_dir+1)*max_acc-0.5*(_dir-1)*max_acc;
}

void LspbTrajPlanner::SetNoTimeLimit(double h)
{
	_t0 = 0;
	if (_maxvel_reached)
	{
		_ta = _vmax/_amax;
		_tf = h/_vmax+_vmax/_amax;
	}
	else
	{
		_ta = sqrt(h/_amax);
		_tf = 2*_ta;
		_vmax = h/_ta;
	}
	_td = _ta;
}

void LspbTrajPlanner::SetTimeLimit(double h, Vector2d duration)
{
	_t0 = duration(0);
	_tf = duration(1);
	double tlen = _tf-_t0;
	if (_maxvel_reached)
	{
		//assert(tlen>=h/_vmax+_vmax/_amax);

		double a = 1;
		double b = -tlen*_amax;
		double c = h*_amax;
		_vmax = (-b-sqrt(pow(b, 2)-4*a*c))/2/a;
		_ta = _vmax/_amax;
	}
	else
	{
		//assert(tlen>2*sqrt(h/_amax));

		_vmax = 2*h/tlen;
		_ta = 0.5*tlen;
		_amax = _vmax/_ta;
	}
	_td = _ta;
}

void LspbTrajPlanner::SetVelConstraint(double h)
{
	_t0 = 0;
	assert(h*_amax>=0.5*fabs(pow(_v0, 2)+pow(_vf, 2)));

	if (h*_amax>pow(_vmax, 2)-0.5*(pow(_v0, 2)+pow(_vf, 2)))
	{
		_maxvel_reached = true;
		_ta = (_vmax-_v0)/_amax;
		_td = (_vmax-_vf)/_amax;
		_tf = h/_vmax+0.5*_vmax/_amax*pow(1-_v0/_vmax, 2)+0.5*_vmax/_amax*pow(1-_vf/_vmax, 2);
	}
	else
	{
		_maxvel_reached = false;
		_vmax = sqrt(h*_amax+0.5*(pow(_v0, 2)+pow(_vf, 2)));
		_ta = (_vmax-_v0)/_amax;
		_td = (_vmax-_vf)/_amax;
		_tf = _ta+_td;
	}
}


RobotTools::JAVP LspbTrajPlanner::GenerateMotion(double t)
{
	RobotTools::JAVP avp;
	if (_maxvel_reached)
	{
		if ((t>=_t0) && (t<=_t0+_ta))
		{
			avp.pos = _q0+_v0*(t-_t0)+0.5*(_vmax-_v0)/_ta*pow(t-_t0, 2);
			avp.vel = _v0+(_vmax-_v0)/_ta*(t-_t0);
			avp.acc = _amax;
		}
		else if ((t>_t0+_ta) && (t<=(_tf-_td)))
		{
			avp.pos = _q0+0.5*_v0*_ta+_vmax*(t-_t0-0.5*_ta);
			avp.vel = _vmax;
			avp.acc = 0;
		}
		else if ((t>(_tf-_td)) && (t<=_tf))
		{
			avp.pos = _qf-_vf*(_tf-t)-0.5*(_vmax-_vf)/_td*pow(_tf-t, 2);
			avp.vel = _vf+(_vmax-_vf)/_td*(_tf-t);
			avp.acc = -_amax;
		}
	}
	else
	{
		if (fabs(_tf-_t0)<EPS)
		{
			avp.pos = _q0;
			avp.vel = 0;
			avp.acc = 0;
		}
		else
		{
			if ((t>=_t0)&&(t<=_t0+_ta))
			{
				avp.pos = _q0+_v0*(t-_t0)+0.5*(_vmax-_v0)/_ta*pow(t-_t0, 2);
				avp.vel = _v0+(_vmax-_v0)/_ta*(t-_t0);
				avp.acc = _amax;
			}
			else if ((t>_tf-_td)&&(t<=_tf))
			{
				avp.pos = _qf-_vf*(_tf-t)-0.5*(_vmax-_vf)/_td*pow(_tf-t, 2);
				avp.vel = _vf+(_vmax-_vf)/_td*(_tf-t);
				avp.acc = -_amax;
			}
		}
	}
	avp.pos *= _dir;
	avp.vel *= _dir;
	avp.acc *= _dir;

	return avp;
}

double LspbTrajPlanner::GetFinalTime()
{
	return _tf;
}

double LspbTrajPlanner::GetDuratoin()
{
	return (_tf-_t0);
}
